src/queue.c

```diff
@@ -1,40 +1,82 @@
 #include "queue.h"
+#include "util.h"
 
-// Every message in the queue is prepended by the message length (ML)
-// ML is sizeof(size_t) in size
-// Tail points to the message data
+#if defined(__GNUC__) || defined(__clang__)
+#define MG_MEMORY_BARRIER() __sync_synchronize()
+#elif defined(_MSC_VER) && _MSC_VER >= 1700
+#define MG_MEMORY_BARRIER() MemoryBarrier()
+#elif !defined(MG_MEMORY_BARRIER)
+#define MG_MEMORY_BARRIER()
+#endif
+
+// Every message in a queue is prepended by a 32-bit message length (ML).
+// If ML is 0, then it is the end, and reader must wrap to the beginning.
+//
+//  Queue when q->tail <= q->head:
+//  |----- free -----| ML | message1 | ML | message2 |  ----- free ------|
+//  ^                ^                               ^                   ^
+// buf              tail                            head                len
 //
-//    |------| ML | message1 |  ML | message2 |--- free space ---|
-//    ^      ^                                ^                  ^
-//   buf    tail                             head               len
-
-size_t mg_queue_space(struct mg_queue *q, char **buf) {
-  size_t ofs;
-  if (q->head > 0 && q->tail >= q->head) {  // All messages read?
-    q->head = 0;                            // Yes. Reset head first
-    q->tail = 0;                            // Now reset the tail
+//  Queue when q->tail > q->head:
+//  | ML | message2 |----- free ------| ML | message1 | 0 |---- free ----|
+//  ^               ^                 ^                                  ^
+// buf             head              tail                               len
+
+void mg_queue_init(struct mg_queue *q, char *buf, size_t size) {
+  q->size = size;
+  q->buf = buf;
+  q->head = q->tail = 0;
+}
+
+static size_t mg_queue_read_len(struct mg_queue *q) {
+  uint32_t n = 0;
+  MG_MEMORY_BARRIER();
+  memcpy(&n, q->buf + q->tail, sizeof(n));
+  assert(q->tail + n + sizeof(n) <= q->size);
+  return n;
+}
+
+static void mg_queue_write_len(struct mg_queue *q, size_t len) {
+  uint32_t n = (uint32_t) len;
+  memcpy(q->buf + q->head, &n, sizeof(n));
+  MG_MEMORY_BARRIER();
+}
+
+size_t mg_queue_book(struct mg_queue *q, char **buf, size_t len) {
+  size_t space = 0, hs = sizeof(uint32_t) * 2;  // *2 is for the 0 marker
+  if (q->head >= q->tail && q->head + len + hs <= q->size) {
+    space = q->size - q->head - hs;  // There is enough space
+  } else if (q->head >= q->tail && q->tail > hs) {
+    mg_queue_write_len(q, 0);  // Not enough space ahead
+    q->head = 0;               // Wrap head to the beginning
   }
-  ofs = q->head + sizeof(size_t);
-  if (buf != NULL) *buf = q->buf + ofs;
-  return ofs > q->len ? 0 : q->len - ofs;
+  if (q->head + hs + len < q->tail) space = q->tail - q->head - hs;
```

Comment:
`else if` ?

Reply:
Nope! On line 51, we wrap head, which can make the condition on line 53 become true.
If we make `else if`, we won't hit that path.